java/sage/plugin/CorePluginManager.java

```diff
@@ -2315,6 +2192,268 @@ public static File getDevPluginsDir() {
+   *
+   * @param pkg Plugin Package
+   * @param localExtract Local File
+   * @param destDir Destination Directory
+   * @param entry Installable Item form which the get the Name and Input Stream
+   * @param priorStateIndex
+   * @param myInstallIndex
+   * @param stagedRenameWriter writer for writing staged renames
+   * @param buf copy data buffer
+   * @return "OK" on success, or and error message if it failed
+   * @throws IOException for any IO related errors
+   */
+  String installFile(PluginWrapper.Package pkg, File localExtract, File destDir, InstallableItem entry, final int priorStateIndex, final int myInstallIndex, StagedRenameWriter stagedRenameWriter, final byte buf[]) throws IOException
+  {
+    if (pkg.overwrite || !localExtract.isFile())
+    {
+      boolean openFailed = false;
+      java.io.File orgLocalExtract = localExtract;
+      // Check to see if the local file is one that we know we shouldn't overwrite while we're running
+      if (localExtract.isFile() && !canModifyPath(localExtract))
+      {
+        openFailed = true;
+      }
+      String orgMD5 = null;
+      if (Sage.DBG) System.out.println("Current extraction local target=" + localExtract);
+      if (localExtract.isFile())
+      {
+        if (pendingFilesystem.containsKey(localExtract))
+        {
+          orgMD5 = (String) pendingFilesystem.get(localExtract);
+          if ("0".equals(orgMD5))
+          {
+            orgMD5 = null;
+          } else
+          {
+            trackFile(localExtract, priorStateIndex, orgMD5, true);
+          }
+        } else
+        {
+          orgMD5 = IOUtils.calcMD5(localExtract);
+          // Update the filetracker with the apriori state of this file
+          // This will also store a backup of the file if needed
+          trackFile(localExtract, priorStateIndex, orgMD5, true);
+        }
+      } else
+      {
+        java.io.File dirParent = localExtract.getParentFile();
+        if (!dirParent.isDirectory() && !dirParent.mkdirs())
+        {
+          if (Sage.DBG) System.out.println("ERROR Cannot create directory from zip archive: " + dirParent);
+          progressMsg = "";
+          return "FAILED - Cannot Create " + dirParent;
+        }
+        if (Sage.DBG)
+          System.out.println("Ensured existence of parent directory " + dirParent + " isDir=" + dirParent.isDirectory());
+        // Also track any parent directories that were automatically created
+        String dirName = entry.getName();
+        int dirIdx = dirName.lastIndexOf(java.io.File.separatorChar);
+        while (dirIdx != -1)
+        {
+          dirName = dirName.substring(0, dirIdx);
+          trackFile(new java.io.File(destDir, dirName), myInstallIndex, "0", false);
+          dirIdx = dirName.lastIndexOf(java.io.File.separatorChar);
+        }
+      }
+
+      FileOutputStream outStream=null;
+      if (!openFailed)
+      {
+        try
+        {
+          outStream = new java.io.FileOutputStream(localExtract);
```

Comment:
Good catch... I'll add a try finally around the block to ensure the outStream gets closed.